Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -531,117 +522,144 @@ section ModelsWithJ
+  convert (hu.out.pow 6).neg
+  norm_num1
 
-lemma ofJ1728_j [Invertible (2 : R)] : (ofJ1728 R).j = 1728 := by
-  field_simp [j, ofJ1728, @val_unitOfInvertible _ _ _ <| invertibleNeg _,
-    WeierstrassCurve.ofJ1728_c₄]
+-- TODO: change to `[IsUnit ...]` once #17458 is merged
+lemma ofJ1728_j [Fact (IsUnit (2 : R))] : (ofJ1728 R).j = 1728 := by
+  field_simp [j, ofJ1728_c₄, ofJ1728_Δ]
   norm_num1
 
 variable {R}
 
-/-- When j and j - 1728 are both invertible,
+-- TODO: change to `[IsUnit ...]` once #17458 is merged
+/-- When j and j - 1728 are both units,
 $Y^2 + (j - 1728)XY = X^3 - 36(j - 1728)^3X - (j - 1728)^5$ is an elliptic curve.
-It is of j-invariant j (see `EllipticCurve.ofJ'_j`). -/
-def ofJ' (j : R) [Invertible j] [Invertible (j - 1728)] : EllipticCurve R :=
-  have := invertibleMul (j ^ 2) ((j - 1728) ^ 9)
-  ⟨WeierstrassCurve.ofJ j, unitOfInvertible <| j ^ 2 * (j - 1728) ^ 9,
-    (WeierstrassCurve.ofJ_Δ j).symm⟩
-
-lemma ofJ'_j (j : R) [Invertible j] [Invertible (j - 1728)] : (ofJ' j).j = j := by
-  field_simp [EllipticCurve.j, ofJ', @val_unitOfInvertible _ _ _ <| invertibleMul ..,
-    WeierstrassCurve.ofJ_c₄]
+It is of j-invariant j (see `WeierstrassCurve.ofJNe0Or1728_j`). -/
+instance IsElliptic.ofJNe0Or1728 (j : R) [h1 : Fact (IsUnit j)] [h2 : Fact (IsUnit (j - 1728))] :
+    (ofJNe0Or1728 j).IsElliptic := by
+  rw [isElliptic_iff, ofJNe0Or1728_Δ j]
+  exact (h1.out.pow 2).mul (h2.out.pow 9)
+
+-- TODO: change to `[IsUnit ...]` once #17458 is merged
+lemma ofJNe0Or1728_j (j : R) [Fact (IsUnit j)] [Fact (IsUnit (j - 1728))] :
+    (ofJNe0Or1728 j).j = j := by
+  field_simp [WeierstrassCurve.j, ofJNe0Or1728_c₄, ofJNe0Or1728_Δ]
   ring1
 
 variable {F : Type u} [Field F] (j : F)
 
-private lemma two_or_three_ne_zero : (2 : F) ≠ 0 ∨ (3 : F) ≠ 0 :=
-  ne_zero_or_ne_zero_of_nat_coprime <| by decide
-
 variable [DecidableEq F]
 
 /-- For any element j of a field $F$, there exists an elliptic curve over $F$
-with j-invariant equal to j (see `EllipticCurve.ofJ_j`).
-Its coefficients are given explicitly (see `EllipticCurve.ofJ0`, `EllipticCurve.ofJ1728`
-and `EllipticCurve.ofJ'`). -/
-def ofJ : EllipticCurve F :=
-  if h0 : j = 0 then
-    if h3 : (3 : F) = 0 then @ofJ1728 _ _ <| invertibleOfNonzero <|
-      two_or_three_ne_zero.neg_resolve_right h3
-    else @ofJ0 _ _ <| invertibleOfNonzero h3
-  else if h1728 : j = 1728 then
-    @ofJ1728 _ _ <| invertibleOfNonzero fun h => h0 <|
-    by rw [h1728, show (1728 : F) = 2 * 864 by norm_num1, h, zero_mul]
-  else @ofJ' _ _ j (invertibleOfNonzero h0) (invertibleOfNonzero <| sub_ne_zero_of_ne h1728)
+with j-invariant equal to j (see `WeierstrassCurve.ofJ_j`).
+Its coefficients are given explicitly (see `WeierstrassCurve.ofJ0`, `WeierstrassCurve.ofJ1728`
+and `WeierstrassCurve.ofJNe0Or1728`). -/
+def ofJ : WeierstrassCurve F :=
+  if j = 0 then
+    if (3 : F) = 0 then
+      ofJ1728 F
+    else
+      ofJ0 F
+  else if j = 1728 then
+    ofJ1728 F
+  else
+    ofJNe0Or1728 j
```

Comment:
These and the lemmas below can be stated in fewer lines, since we no longer need `invertibleOfNonzero`.

Reply:
Yeah that's what I'm thinking. The statements of the other lemmas can each be one line.

Reply:
Done.